Initial  support for sparse complex tensors constructors  for CPU/CUDA

aocsa opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55739 sparse complex: add linear algebra, addmm
* **#54153 Initial  support for sparse complex tensors constructors  for CPU/CUDA**

Currently, sparse tensors only support real floating point tensors. Complex support is added in this PR for CPU/CUDA.

- [x] add complex support (torch.cfloat and torch.cdouble) to torch.sparse_coo_tensor constructors
- [x] add complex support to coalesce function
- [x] add complex support to to_dense function
- [x] add complex support to to_sparse function
- [x] add complex support to sparse_add function
- [x] add unit tests

Note: This PR contains only complex support for torch.sparse_coo_tensor *forward/backward* function and the related ops used with this function (coalesce, to_dense, to_sparse, and sparse_add). The following PRs in ghstack should cover other sparse operations to have a more complex sparse support, specifically related with the use of specific APIs for accelerated linear algebra.

Note: Before using ghstack the original PR  was  #50984

Differential Revision: [D27765618](https://our.internmc.facebook.com/intern/diff/D27765618)

aocsa committed:
Initial  support for sparse complex tensors constructors  for CPU/CUDA

[ghstack-poisoned]

mruberry commented on test/test_autograd.py:
```diff
@@ -6861,7 +6821,8 @@ def test_parameter_resize(self, device):
             m.sum().backward()
 
-    def test_sparse_ctor_getter_backward(self, device):
+    @dtypes(torch.double, torch.cdouble)
```
Why double and cdouble (vs float and cfloat, for example)?

anjali411 replied:
It's because gradcheck for `torch.cfloat` and `torch.float` with default eps/tolerance will fail.

mruberry commented on test/test_autograd.py:
```diff
@@ -6895,6 +6856,49 @@ def fn(v):
+            [0, 3, 4],
+            [0, 2, 2],
+        ], dtype=torch.long)
+        v1 = torch.tensor([[1, 2], [4, 5], [7, 8]], dtype=dtype)
```
use make_tensor(), this test currently ignores device, make_tensor() helps avoid that issue (while providing other features) by requiring device and dtype be specified

mruberry commented on test/test_autograd.py:
```diff
@@ -6895,6 +6856,49 @@ def fn(v):
+
+    @dtypes(*torch.testing.floating_and_complex_types())
+    def test_sparse_backward(self, device, dtype):
+        class FixedGradientFunction(Function):
```
Would you comment on the design of this function? What is it intended to validate, and how does it use FixedGradientFunction?

anjali411 replied:
yeah and also I think we should use `gradcheck` in this test

aocsa replied:
Note that this test is not new, is the same as the previous one. What I understand this test check `sparse_backward` for different cases (`dense + sparse` cases) and IMO it's clear using this way.

mruberry commented on torch/testing/_internal/common_device_type.py:
```diff
@@ -484,42 +484,48 @@ def instantiate_device_type_tests(generic_test_class, scope, except_for=None, on
-        device_type_test_class.__module__ = generic_test_class.__module__
-        scope[class_name] = device_type_test_class
+
+        for is_uncoalesced in set([add_uncoalesced_tests, False]):
```
I don't think we want to edit the device generic test framework like this to support toggling coalescing. 

What does this want to do, exactly? Is it that for a test template "test_foo" it wants to instantiate a cross-product of device x dtype x coalesced/uncoalesced, so an instantiated test name would look like "test_foo_cpu_float32_coalesced"?

aocsa replied:
It only instantiates when  `add_uncoalesced_tests` is True if not it remains with the same behavior. 

So, when `add_uncoalesced_tests` is True it generates also the uncoalesced version.  
So in the example two versions will be instantiated:
- "test_foo_cpu_float32"
- "test_foo_cpu_float32_uncoalesced"

mruberry replied:
That makes sense and it's a great goal.

What about adding a decorator, like @coalescedonoff that modifies test generation here:

https://github.com/pytorch/pytorch/blob/789dc6d445136307070486fccf6c9276a722a384/torch/testing/_internal/common_device_type.py#L327

And it could pass it a "is_coalesced" kwarg to the instantiated test function, so the function template signature would become:

```
test_foo(self, device, dtype, *, is_coalesced)
```

and the instantiated functions would have names like test_foo_cuda_float32_coalesced and test_foo_cpu_cfloat_uncoalesced?

A more general approach would be to create a "variants" decorator. It might look like this:

```
@variants(
  ("coalesced", {"is_coalesced : True"}),
  ("uncoalesced", {"is_coalesced : False"})
)
```

That might look like a lot, but with partial this pattern could be saved under a simpler name like @coalescedonoff. Each entry in variants could specify how to extend the test's name, then provide the dict of kwargs and values to pass to the instantiated test.

This is a little like pytest's parameterize functionality, although they support many more options. If you'd like to we could look at some alternative designs based on parameterize's UX.

Would an approach like this make sense? Or should we do something else?

aocsa replied:
Thanks @mruberry, In this PR I was using `is_uncoalesced` as an object attribute for each instance as most of the test methods requires both versions, but looking it in detail, there are some test that don't require uncoalesced tensors. 
So, I agree with this approach.  I think in the meanwhile the @coalescedonoff is enough. I will add this change in a new PR #54572 as it was suggested.

mruberry replied:
SGTM!

mruberry replied:
Note that one other alternative is to do something like this:

https://github.com/pytorch/pytorch/blob/8c798e062216278673a75bac0848ea69a8bd3f03/torch/testing/_internal/common_cuda.py#L102

Here a new test isn't generated, but the current test is run twice. This is a little hacky, however.

mruberry reviewed (changes requested):
Hey @aocsa! There's a lot of good stuff in this PR. It not only adds support for creating complex sparse tensors, but moves the sparse test suite to the device generic test framework, and proposes a significant extension to it, too!

Are these changes separable? For example, would it be helpful to focus on the device generic test framework port first? I'd like to review that design with you more.

anjali411 commented on test/expect/TestSparseCPU.test_print_cpu_float64.expect:
```diff
@@ -46,25 +46,23 @@ tensor(indices=tensor([], size=(0, 10)),
 # sparse tensor
 tensor(indices=tensor([], size=(0, 10)),
        values=tensor([], size=(10, 0)),
-       size=(0,), nnz=10, dtype=torch.float32, layout=torch.sparse_coo)
```
is the default dtype set to `double`?

aocsa replied:
Yes the default is dtype is double.   I will update the changes related to the tests in a new PR #54572

anjali411 commented on test/test_sparse.py:
```diff
@@ -2364,10 +2398,10 @@ def test_factory_device_type_inference(self):
-    @cpu_only
-    def test_factory_copy(self):
+    @onlyCPU
+    def test_factory_copy(self, device):
```
why aren't we running this test for complex types?

aocsa replied:
I added test cases for complex data

anjali411 commented:
@aocsa this PR looks very good overall. I agree with @mruberry that it'd be easy to ,both, review and land these changes if this PR were split into two PRs -- one which is more focused on adding device generic tests and then the following PR which adds complex support and leverages the device generic tests to test complex support.

aocsa commented:
As suggested by @anjali411 and @mruberry  a different PR  #54572  was created, adding device generic tests to `test_sparse`. This PR will be looked first.

aocsa commented on test/expect/TestSparseCUDA.test_print_cuda_float64.expect:
```diff
@@ -135,21 +134,20 @@ tensor([[0, 1, 2]], device='cuda:0')
                       [0.6667, 0.8889, 1.1111],
                       [1.3333, 1.5556, 1.7778]]),
-       device='cuda:0', size=(100, 3), nnz=3, dtype=torch.float32,
-       layout=torch.sparse_coo, requires_grad=True)
```
the `test_print`  is tested for both `torch.float` and `torch.double`. So that part if for float32

aocsa committed:
Update on "Initial  support for sparse complex tensors constructors  for CPU/CUDA"


Before using ghstack the original PR  was  #50984


[ghstack-poisoned]

aocsa committed:
Update on "Initial  support for sparse complex tensors constructors  for CPU/CUDA"


Before using ghstack the original PR  was  #50984


[ghstack-poisoned]

aocsa committed:
Update on "Initial  support for sparse complex tensors constructors  for CPU/CUDA"


Before using ghstack the original PR  was  #50984


[ghstack-poisoned]

aocsa committed:
Update on "Initial  support for sparse complex tensors constructors  for CPU/CUDA"


Before using ghstack the original PR  was  #50984


[ghstack-poisoned]

aocsa committed:
Update on "Initial  support for sparse complex tensors constructors  for CPU/CUDA"


Before using ghstack the original PR  was  #50984


[ghstack-poisoned]

aocsa committed:
Update on "Initial  support for sparse complex tensors constructors  for CPU/CUDA"


Before using ghstack the original PR  was  #50984


[ghstack-poisoned]

aocsa commented:
I rebased this PR on top of the latest changes,  including PR #54572  (adding device generic tests to test_sparse)  and #52874 (memory leak fixed for reference cycle in sparse coalesce graphs) derived from the original PR. Moreover, I polished a few things and addressed the latest feedbacks. 
This contains only complex support for `torch.sparse_coo_tensor` constructor function and the related ops used with this function (`coalesce, to_dense, to_sparse, and sparse_add`). I want to use this PR as a baseline  so the following PRs in ghstack should cover other sparse operations to have a more complex sparse support, specifically related with the use of specific APIs for accelerated linear algebra.

I think it is ready for review / merging, let me know what you think. 

cc @mruberry, @anjali411

aocsa committed:
Update on "Initial  support for sparse complex tensors constructors  for CPU/CUDA"


Currently, sparse tensors only support real floating point tensors. Complex support is added in this PR for CPU/CUDA.

- [x] add complex support (torch.cfloat and torch.cdouble) to torch.sparse_coo_tensor constructors
- [x] add complex support to coalesce function
- [x] add complex support to to_dense function
- [x] add complex support to to_sparse function
- [x] add complex support to sparse_add function
- [x] add unit tests

Note: This PR contains only complex support for torch.sparse_coo_tensor fordward function and the related ops used with this function (coalesce, to_dense, to_sparse, and sparse_add). The following PRs in ghstack should cover other sparse operations to have a more complex sparse support, specifically related with the use of specific APIs for accelerated linear algebra.

Note: Before using ghstack the original PR  was  #50984


[ghstack-poisoned]

ezyang commented:
Looks like there are a bunch of build failures

aocsa committed:
Update on "Initial  support for sparse complex tensors constructors  for CPU/CUDA"


Currently, sparse tensors only support real floating point tensors. Complex support is added in this PR for CPU/CUDA.

- [x] add complex support (torch.cfloat and torch.cdouble) to torch.sparse_coo_tensor constructors
- [x] add complex support to coalesce function
- [x] add complex support to to_dense function
- [x] add complex support to to_sparse function
- [x] add complex support to sparse_add function
- [x] add unit tests

Note: This PR contains only complex support for torch.sparse_coo_tensor fordward function and the related ops used with this function (coalesce, to_dense, to_sparse, and sparse_add). The following PRs in ghstack should cover other sparse operations to have a more complex sparse support, specifically related with the use of specific APIs for accelerated linear algebra.

Note: Before using ghstack the original PR  was  #50984

Differential Revision: [D27765618](https://our.internmc.facebook.com/intern/diff/D27765618)

[ghstack-poisoned]

aocsa committed:
Update on "Initial  support for sparse complex tensors constructors  for CPU/CUDA"


Currently, sparse tensors only support real floating point tensors. Complex support is added in this PR for CPU/CUDA.

- [x] add complex support (torch.cfloat and torch.cdouble) to torch.sparse_coo_tensor constructors
- [x] add complex support to coalesce function
- [x] add complex support to to_dense function
- [x] add complex support to to_sparse function
- [x] add complex support to sparse_add function
- [x] add unit tests

Note: This PR contains only complex support for torch.sparse_coo_tensor fordward function and the related ops used with this function (coalesce, to_dense, to_sparse, and sparse_add). The following PRs in ghstack should cover other sparse operations to have a more complex sparse support, specifically related with the use of specific APIs for accelerated linear algebra.

Note: Before using ghstack the original PR  was  #50984

Differential Revision: [D27765618](https://our.internmc.facebook.com/intern/diff/D27765618)

[ghstack-poisoned]

aocsa committed:
Update on "Initial  support for sparse complex tensors constructors  for CPU/CUDA"


Currently, sparse tensors only support real floating point tensors. Complex support is added in this PR for CPU/CUDA.

- [x] add complex support (torch.cfloat and torch.cdouble) to torch.sparse_coo_tensor constructors
- [x] add complex support to coalesce function
- [x] add complex support to to_dense function
- [x] add complex support to to_sparse function
- [x] add complex support to sparse_add function
- [x] add unit tests

Note: This PR contains only complex support for torch.sparse_coo_tensor fordward function and the related ops used with this function (coalesce, to_dense, to_sparse, and sparse_add). The following PRs in ghstack should cover other sparse operations to have a more complex sparse support, specifically related with the use of specific APIs for accelerated linear algebra.

Note: Before using ghstack the original PR  was  #50984

Differential Revision: [D27765618](https://our.internmc.facebook.com/intern/diff/D27765618)

[ghstack-poisoned]

aocsa committed:
Update on "Initial  support for sparse complex tensors constructors  for CPU/CUDA"


Currently, sparse tensors only support real floating point tensors. Complex support is added in this PR for CPU/CUDA.

- [x] add complex support (torch.cfloat and torch.cdouble) to torch.sparse_coo_tensor constructors
- [x] add complex support to coalesce function
- [x] add complex support to to_dense function
- [x] add complex support to to_sparse function
- [x] add complex support to sparse_add function
- [x] add unit tests

Note: This PR contains only complex support for torch.sparse_coo_tensor *forward/backward* function and the related ops used with this function (coalesce, to_dense, to_sparse, and sparse_add). The following PRs in ghstack should cover other sparse operations to have a more complex sparse support, specifically related with the use of specific APIs for accelerated linear algebra.

Note: Before using ghstack the original PR  was  #50984

Differential Revision: [D27765618](https://our.internmc.facebook.com/intern/diff/D27765618)

[ghstack-poisoned]

ezyang commented:
@mruberry are you going to take another go at this?

mruberry commented:
> @mruberry are you going to take another go at this?

Thanks for the ping, @ezyang. I missed this notification. I'll try to take a look tomorrow.

mruberry reviewed (approved):
Cool! This looks good to me now. @anjali411, would you like to take another look?

mruberry commented:
Unfortunately this appears to be breaking pytorch_windows_vs2019_py36_cpu_test1, so I had to unland it. See https://app.circleci.com/pipelines/github/pytorch/pytorch/311390/workflows/6ba089d9-ab02-4c39-a41c-f174e60b7d77/jobs/12913071.
